dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/JCloudsAppStorageService.java

```diff
@@ -338,10 +339,17 @@ public App installApp( File file, String filename, Cache<App> appCache )
         try ( ZipFile zip = new ZipFile( file ) )
         {
             // -----------------------------------------------------------------
-            // Parse ZIP file and it's manifest.webapp file.
+            // Determine top-level directory name, if the zip file contains one
             // -----------------------------------------------------------------
 
-            ZipEntry entry = zip.getEntry( MANIFEST_FILENAME );
+            String prefix = ZipFileUtils.getTopLevelDirectory( zip.entries().asIterator() );
+            log.debug( "Detected top-level directory '" + prefix + "' in zip" );
```

Comment:
Very minor, you can do this with slf4j to avoid string concatenation:

> log.debug( "Detected top-level directory '{}' in zip archive", prefix );

Reply:
Good to know, thanks @larshelge !  I was copying the style used elsewhere in this file, slf4j is definitely nicer!